### ISSUE TITLE
facter > 3.1.8 fails when used in ansible

### ISSUE THREAD

**Issue by merlin-tc** [ansible/ansible#23695]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
modules/system/facter.py

##### ANSIBLE VERSION
ansible 2.3.0.0

##### CONFIGURATION
standard ansible configuration

##### OS / ENVIRONMENT
Gentoo Linux

##### SUMMARY
Upstream facter removed the --puppet flag which causes facter to fail silently even though it's called by ansible.
If called directly it shows the error.
Removing "--puppet" from the facter command fixes the problem if the facter module is called directly. Strangely enough there is still no output if the ansible "setup" module is used, which should pull in the facter output as well.

Output from older version of facter where it still works:
  -p [ --puppet ]                (Deprecated: use `puppet facts` instead) Load 
                                 the Puppet libraries, thus allowing Facter to 
                                 load Puppet-specific facts.

##### STEPS TO REPRODUCE

```yaml
ansible localhost -m facter
```

##### EXPECTED RESULTS
Add facts gathered by ansible.

##### ACTUAL RESULTS
```yaml
META: ran handlers
Using module file /usr/lib64/python2.7/site-packages/ansible/modules/system/facter.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: merlintc
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python2.7 && sleep 0'
localhost | FAILED! => {
    "changed": false, 
    "cmd": "/usr/bin/facter --puppet --json", 
    "failed": true, 
    "invocation": {
        "module_args": {}
    }, 
    "msg": "2017-04-18 13:04:24.366258 FATAL puppetlabs.facter - unhandled exception: boost::too_many_args: format-string referred to fewer arguments than were passed", 
    "rc": 1, 
    "stderr": "2017-04-18 13:04:24.366258 FATAL puppetlabs.facter - unhandled exception: boost::too_many_args: format-string referred to fewer arguments than were passed\n", 
    "stderr_lines": [
        "2017-04-18 13:04:24.366258 FATAL puppetlabs.facter - unhandled exception: boost::too_many_args: format-string referred to fewer arguments than were passed"
    ], 
    "stdout": "", 
    "stdout_lines": []
}
```
```

```

**Comment by alikins** [comment]
@merlin-tc which facter version are you seeing this with?

looks like 3.0.0 removed '--puppet' and '3.0.2' added it back (https://docs.puppet.com/facter/3.0/release_notes.html#facter--p-restored)

**Comment by alikins** [comment]
oops, I missed the version info in the description (> 3.1.8 )

I'll take a look and see if we can support old and new cli.

**Comment by merlin-tc** [comment]
Removing the --puppet flag from facter does **NOT** fix the problem. It silences the error but the output of facter is not used by ansible.

**Comment by merlin-tc** [comment]
There is another file still being called with -puppet.
We need to patch "ansible/module_utils/facts.py" and remove the "--puppet" flag there. Only after that change I got the all the facter variables.

**Comment by nrwahl2** [comment]
@alikins @merlin-tc : I'm not able to reproduce the issue. I'm running Ansible 2.4 and Facter 3.6.5. It works as expected when running both `-m facter` and `-m setup -a 'filter=*facter*'`.

It looks like the backward-compatibility issues with `facter` were fixed in 3.1.1, and I see no notes after 3.1.8 describing any changes to `facter -p`. https://docs.puppet.com/facter/3.1/release_notes.html#regression-fix-running-facter---puppet-produces-warning-and-debug-messages

Edit: Whoops! Just saw the merged PR above. Not sure what issue merlin-tc is facing. Mine works the same with or without the `--puppet` flag.

```
$ grep '\-\-puppet' facter.py
        rc, out, err = module.run_command(facter_path + " --puppet --json")
$ ansible localhost -i ~/hosts.ini -m setup -a 'filter=*facter*' | wc -l
388
$ vim facter.py
$ grep '\-\-puppet' facter.py
$ ansible localhost -i ~/hosts.ini -m setup -a 'filter=*facter*' | wc -l
387
```

**Comment by nrwahl2** [comment]
@merlin-tc , could you try with the devel branch of Ansible?

**Comment by merlin-tc** [comment]
Fixed for me in 2.4.
Thank you very much!
